blocks.md

```diff
@@ -0,0 +1,80 @@
+
+<h4>2. Block Plugins are separate blocks.</h4>
+
+Block plugins are intended to be single purpose, separate, independent blocks, not bundles or compendiums of many blocks. In most cases, a Block Plugin should contain only one (single) block. It should not contain multiple blocks that could reasonably be split up into separate, independent plugins.
+
+Block plugins may contain more than one block where a clearly necessary parent/child dependency exists; for example a list block that contains list item blocks.
+
+<h4>3. Plugin titles should reflect the block title.</h4>
+
+We want users to interact with blocks, not product names. Block Plugins should be named to reflect the block; for example "Image Grid" block, "Improved Slider", "Business Hours" block. Please avoid plugin names unrelated to the block itself like "Widgets Incorporated Block", and refrain from block names that include trademarks or names that you do not legally represent; for example you cannot name a block "Facerange Block" unless you work for Facerange and have their permission.
+
+<h4>4. Block Plugins must include a <code>block.json</code> file.</h4>
+
+The Block Registration RFC outlines the format of a `block.json` file: https://github.com/WordPress/gutenberg/blob/master/docs/rfc/block-registration.md
+
+Block Plugins must include a valid `block.json` file. In addition to the requirements specified in the RFC, the `block.json` must include the following attributes:
+* name
+* title
+* At least one of: script, editorScript
+* At least one of: style, editorStyle
+
+<h4>5. Block Plugins must not require payment or a paid account to function.</h4>
+
+This includes payment for premium features. You may use the donation link feature.
+
+<h4>6. Block Plugins should work independently.</h4>
+
+Block Plugins must not require another plugin in order to function.
+
+They may use an external/cloud API where necessary, but not if that API requires a paid account for access.
+
+They should not rely on an external API or cloud service for functions that could be performed locally.
+
+<h4>7. Server-side code should be kept to a minimum.</h4>
+
+Since Block Plugins are WordPress plugins, they necessarily contain PHP code for metadata and initialisation. As much as possible, they should avoid including additional PHP code or server-side libraries. The WordPress REST API should be the preferred interface to WordPress, rather than custom server-side code.
+
+We recognise that there are limits to the REST API, and situations where server-side PHP is the only performant way to implement a feature. In those situations, PHP code may be included. Please ensure that such code is clearly written, used as little as possible, and well documented.
+
+<h4>8. Must not promote other blocks, plugins, or themes.</h4>
```

Comment:
How is "promote" defined here? If I have a single-block plugin that's fully functional on its own, but can be _enhanced_ by running it in conjunction with another plugin, is just having that interactivity considered promotional?

Imagine Block Plugin X has a tight set of defined features which will be perfectly suitable for most users. However, if you happen to use it alongside Full-Featured Plugin X, you can enhance it with additional features which are directly related to the functional purpose of Block Plugin X.

Does having that potential interactivity between Block Plugin X and Full-Featured Plugin X mean that Block Plugin X is not meeting this guideline? Or does it depend on how it's presented to the user?

Reply:
While it makes sense to require the downloaded block itself provide features that don’t need any payment, why would a link in the settings/config of that block to learn more about additional premium features (sales of which can help the developer support the free version) not be allowed similar to existing plugins? For example on the bottom of the block’s settings output? Having guidelines around taking over other parts of the UI or not including any admin notices above the editor would make sense, though.

Reply:
I’m curious about this as well. A link or some other way of promotion to an existing plugin should be allowed. Guidelines around promotion should be added instead of just disallowing all of them.

Perhaps adding a uniform note with a link in the block description would work, something like “this block is a part of XYZ blocks” kind of like the donation button

Reply:
Without the ability to create a market around blocks, there will be less blocks. Donations are generally considered to be a non-viable business model and the best way to incentivize more blocks of better quality is to allow for a freemium model in addition to or really replacing the donation link or button.

There's an opportunity here to define up-sell channels for potential use in a way that won't disrupt flow, but can still catch the eye of the user should they be curious for more information. Providing a link or button on the bottom of the block settings or even a small link in the description could be enough to get people interested if there's more features available.

Nothing flashy, nothing overpowering, but the presence of a link or button would be felt without disrupting flow. If everyone had to play by the same rules (ie. defined elements, placement, etc.) then quality becomes an even more important factor.

Without this, the vast majority of people who want to write blocks at a professional level is going to be restricted to developers working either in their spare time for the fun of it or those paid by corporations to help extend their influence. We lose a whole sector of development back into 3rd party marketplaces which are already overflowing with block compilations and bundles, and this entire effort which has serious long-lasting potential for the future of WordPress creation and customization gets stunted.

Reply:
If the goal is to prevent spammy advertisements in the block directory, there should just be very strict guidelines as to what is permitted for promotion instead, and perhaps define what "promote" is.

Reply:
I don't see why this needs to be restricted. People will vote with their feet if a block is overly promotional. Making those rules any different here from the normal plugin directory feels like a mistake.

Don't break the possibility for small blocks to interact with plugins (and thus maybe their need to promote back and forth) for shared functionality etc. Don't break freemium. It's what our ecosystem depends upon.

Reply:
> People will vote with their feet if a block is overly promotional.

This is a special case because of the seamless install. You can install a block plugin, use it in your post, and publish your post without ever leaving the editor. A user could do that, perhaps adding several new block plugins, and only visit the main dashboard hours or days later. We don't want them discovering their dashboard is full of notices and promos long after they've used the block, and when there's no clear causal link between installing a particular block and its side effects outside the editor.

I expect that the guidelines will be relaxed later. This is a very new thing that's the first big change to the way plugins are installed in core for a long time. One bad actor could easily sour it. Let's try to keep it as simple and safe as possible to start with, and open things up over time as we learn from real experience.

Reply:
Doing that will literally limit us in what we can do and make it almost useless to us. So I’d suggest changing your stance. Also, this feels like you’re the one making the final decision, and that decision has already been made. I object to both.

Reply:
Doing that will literally limit us in what we can do and make it almost useless to us. So I’d suggest changing your stance. Also, this feels like you’re the one making the final decision, and that decision has already been made. I object to both.

Reply:
> We don't want them discovering their dashboard is full of notices and promos long after they've used the block

I think things are being likened to the admin notices area where it's possible for things to get out of hand and have a ton of things floating in there. But I don't think this is what people are pointing out here, nor is it what people want to do.

A simple thing like a note in the block's description might suffice for now (visible only on the block that has been installed). Yet, even a small promotion like this isn't allowed.

![Screen Shot 2020-07-13 at 6 48 30 PM](https://user-images.githubusercontent.com/1033611/87296529-7bdb5180-c539-11ea-8653-ad73d23e9649.jpg)

Reply:
Like @tellyworth said, this guideline might be relaxed later, but I think this is a good first step for the directory.

Reply:
It's a good first step if the block directory is only intended for simple, layout-based plugins with no database requirement or enhanced functionality. For example the one [event block](https://wordpress.org/plugins/event/) doesn't actually let you pick from events on a calendar or link to a page with more details, but rather type the basic event details in manually. Under this restriction we can't even include the block in s separate plugin with a link to download the plugin used to actually save and manage events.

Like any plugin in the WordPress directory some validation needs to be done by the user to ensure it will be useful to them. Why arbitrarily restrict what can be included, and make it harder for the user to find and install what they need, as long as the plugin makes a block available and isn't doing something malicious?